MessageCommon.sedona

```diff
@@ -0,0 +1,144 @@
+  **
+  ** reset current message's local state
+  **
+  internal virtual void reset() {
+  }
+
+  **
+  ** publish linked value to MQTT broker
+  **
+  internal virtual void publish(Obj handle)
+  {
+  }
+
+  **
+  ** subscribe topic on MQTT broker refered by given handle
+  **
+  internal virtual void subscribe(Obj handle)
+  {
+  }
+
+  **
+  ** unsubscribe topic on MQTT broker refered by given handle
+  **
+  internal virtual void unsubscribe(Obj handle)
+  {
+  }
+
+  internal virtual void updateFullName()
+  {
+    fullTopicName.copyFromStr(prefixName, fullTopicNameLen)
+    if (prefixName.length() != 0)
+      appendStr(fullTopicName, "/", fullTopicNameLen)
+    appendStr(fullTopicName, topic.toStr(), fullTopicNameLen)
+    validateInputs()
+  }
+
+  protected virtual void updatePrefix(Str prefix)
+  {
+    prefixName.copyFromStr(prefix, fullTopicNameLen)
+    updateFullName()
+  }
+
+
+  static private bool appendStr(Str dst, Str src, int maxDstLen)
+  {
+    bool res = true;
+    int startIndx = dst.length()
+    int bytesCopy = src.length()
+    bytesCopy = bytesCopy + 1
+    if ((dst.length() + bytesCopy) > maxDstLen)
+    {
+      bytesCopy = maxDstLen - dst.length()
+      res = false
+    }
+
+    for(int i = 0; i < bytesCopy; i++)
+      dst.set(startIndx + i, src.get(i))
+
+    return res
+  }
+
+  virtual override void changed(Slot slot)
+  {
+    super.changed(slot)
+    if (slot.name == "topic")
+      updateFullName()
+  }
+
+  protected void validateInputs()
+  {
+    Component parentComponent = Sys.app.lookup(parent)
+    Str parentName = parentComponent?.type?.name
+    if (!parentComponent.type.is(Publisher.type) &&
+        !parentComponent.type.is(Subscriber.type) &&
+        !parentComponent.type.is(TopicGroup.type))
+    {
+      error := Err_InvalidParent;
+    }
+    else if (topic.toStr().length() > 0)
+    {
+      if (parentComponent.type.is(TopicGroup.type))
+        error := ((TopicGroup)parentComponent).error
+      else
+        error := Ok
+    }
+    else
+      error := Err_InvalidTopic
+  }
+
+
+  override int parentEvent(int eType, Component parent)
+  {
+    if (eType == Component.ADDED &&
+	(!parent.type.is(Publisher.type) &&
+	 !parent.type.is(Subscriber.type)&&
+	 !parent.type.is(TopicGroup.type)))
+      error := Err_InvalidParent;
+    return 0
+  }
+
+
+  override virtual void start()
+  {
+    if (fullTopicName.equals(""))
+    {
+      Str prefix = ""
+      Component parentComponent = Sys.app.lookup(parent)
+      if (parentComponent.type.is(TopicGroup.type))
+      {
+        TopicGroup t = (TopicGroup)parentComponent
+        prefix = t.fullTopicName
+      }
+
+      updatePrefix(prefix)
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////
+  // Fields
+  ////////////////////////////////////////////////////////////////
+  define int fullTopicNameLen = 400
+  protected inline Str(fullTopicNameLen) prefixName = ""
```

Comment:
I don't think it's a big problem honestly. Especially because this kit is targeting Linux-based systems, where memory requirements are already pretty high in compare to more embedded stuff like cortex-m. I increased default 64 length for topic name, because if user wants subscribe single topic. It easier to just use single MessageInt component with a full topic path instead of build up whole hierarchy of TopicGroup component. Probably for most cases 64 characters will be enough.

I'm not sure that it's possible to decrease memory usage on the fly in sedona. 
And thanks for fixing my implementation. It's nice to get rid of lastSubscribedTopic. That brings a lot of memory saving.